google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import static com.google.cloud.spanner.SessionImpl.NO_CHANNEL_HINT;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.cloud.spanner.SessionClient.SessionConsumer;
+import com.google.cloud.spanner.SpannerException.ResourceNotFoundException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * {@link DatabaseClient} implementation that uses a single multiplexed session to execute
+ * transactions.
+ */
+final class MultiplexedSessionDatabaseClient extends AbstractMultiplexedSessionDatabaseClient {
+
+  /**
+   * Represents a single transaction on a multiplexed session. This can be both a single-use or
+   * multi-use transaction, and both read/write or read-only transaction. This can be compared to a
+   * 'checked out session' of a pool, except as multiplexed sessions support multiple parallel
+   * transactions, we do not need to actually check out and exclusively reserve a single session for
+   * a transaction. This class therefore only contains context information about the current
+   * transaction, such as the current span, and a reference to the multiplexed session that is used
+   * for the transaction.
+   */
+  static class MultiplexedSessionTransaction extends SessionImpl {
+    private final MultiplexedSessionDatabaseClient client;
+
+    private final boolean singleUse;
```

Comment:
1. How is the channel hint managed for multi-use transactions?
2. Given we will always set the `options` at `SessionImpl`, the random hint logic that was introduced in `AbstractReadContext` becomes dead code. Do we clean it up here itself? Or do that in a separate PR? I'm fine either ways.

Reply:
1. Channel hint for multi-use read-only transactions are always random. In theory we could use the same strategy for them as for single-use read-only transactions, but I think we should not, because it would make the 'channel-hint-bitset' depend on read-only transactions always being closed. Failure to close read-only transactions has been a common source of session leaks in the past, and that is exactly what we want to get rid of with multiplexed sessions. Adding a new dependency on closing all read-only transactions here could in theory give us a bit better latency if someone is running 1qps multi-use read-only transactions, but I don't think the benefit of that outweighs the potential downside.
2. The code for assigning a random hint in `AbstractReadContext` is still used. We only create an `options` instance if the channel hint is __not__ the special value `NO_CHANNEL_HINT`. See https://github.com/googleapis/java-spanner/blob/5979f2de26b9ec0a4c6a5f2c91b27304dcd9404a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionImpl.java#L125. What happens in this case is the following:
  2.1. `MultiplexedSessionDatabaseClient` returns `NO_CHANNEL_HINT` if we should just use a random channel: https://github.com/googleapis/java-spanner/blob/5979f2de26b9ec0a4c6a5f2c91b27304dcd9404a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java#L315
  2.2. That value is then passed to the constructor of `SessionImpl`.
  2.3. `SessionImpl` checks if the channel hint is `NO_CHANNEL_HINT`, and if so does not create an `options` instance, but instead calls `sessionReference.getOptions()`.
  2.4. `sessionReference.getOptions()` returns null for multiplexed sessions and the affiliated channel for regular sessions.